Fix #26840 -- Move `teardown_databases` to `django.test.utils`.

pelme opened the pull request:
Moved `teardown_databases` from a method on DiscoverRunner to a utility function. To be consistent with other testing utilities, this patch also moves `setup_databases` into `django.test.utils.`.

This commit makes it easier to build custom test runners without having to rely on tricks with DiscoverRunner, i.e.:
https://github.com/pytest-dev/pytest-django/blob/def28820dd24283ee32650f2adb8ace21d05a0a1/pytest_django/fixtures.py#L58-L59

carljm commented:
Looks good to me! Normally we require a Trac ticket to accompany a change of this size; I've created https://code.djangoproject.com/ticket/26840

Judging by past experience, @timgraham may have comments on this ;-) once you've passed his review, you'll want to squash this down to a single commit that observes our standard commit message format.

pelme commented:
Thank you @carljm! 👍

I have squashed the commits and updated the commit message to include the ticket number.

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -230,7 +230,13 @@ Templates
+  easier to build custom test runners. :func:`~django.test.utils.teardown_databases` was previously only available on
+  the :class:`~django.test.runner.DiscoverRunner` as a method.
+
+  To be consistent with the :func:`~django.test.utils.teardown_databases` and
```
This goes under "Features deprecated" not "Minor Features" -- see https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature for other deprecation requirements.

timgraham commented on docs/topics/testing/advanced.txt:
```diff
@@ -610,6 +604,22 @@ utility methods in the ``django.test.utils`` module.
     services.
 
+.. function:: setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, parallel=0, **kwargs)
+
```
`.. versionadded:: 1.11`

timgraham commented on django/test/utils.py:
```diff
@@ -137,6 +139,156 @@ def teardown_test_environment():
 
+def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, parallel=0, **kwargs):
+    """
+    Creates the test databases.
```
"Create" (use PEP257 verb-style for new code)

timgraham commented on docs/topics/testing/advanced.txt:
```diff
@@ -554,27 +554,21 @@ Methods
     Returns the result produced by the running the test suite.
 
-.. method:: DiscoverRunner.teardown_databases(old_config, **kwargs)
+.. method:: DiscoverRunner.setup_databases()
```
was removing `**kwargs` intentional?

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -230,7 +230,13 @@ Templates
 ~~~~~
 
-* ...
+* :func:`~django.test.utils.teardown_databases` has been introduced to make it
```
The new teardown_databases function makes it easier to...

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -230,7 +230,13 @@ Templates
 
-* ...
+* :func:`~django.test.utils.teardown_databases` has been introduced to make it
+  easier to build custom test runners. :func:`~django.test.utils.teardown_databases` was previously only available on
```
wrap doc lines at 79 characters

pelme commented:
Thanks for the review @timgraham.

I've addressed your comments and followed the deprecation guideline to the best of my knowledge. The commit is rebased and squashed to the latest master!

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -403,3 +403,12 @@ Miscellaneous
   :class:`~django.contrib.auth.views.PasswordResetConfirmView`, and
   :class:`~django.contrib.auth.views.PasswordResetCompleteView`.
+
+* The new :func:`~django.test.utils.teardown_databases` makes it easier to
```
This should be more explicit about what's deprecated. Also a note in `docs/internals/deprecation.txt` is missing.

timgraham commented on django/test/utils.py:
```diff
@@ -141,6 +143,156 @@ def teardown_test_environment():
+    ordered_test_databases = []
+    resolved_databases = set()
+
+    # Maps db signature to dependencies of all it's aliases
```
its

timgraham commented on django/test/utils.py:
```diff
@@ -141,6 +143,156 @@ def teardown_test_environment():
+        changed = False
+        deferred = []
+
+        # Try to find a DB that has all it's dependencies met
```
its

timgraham commented on django/test/utils.py:
```diff
@@ -141,6 +143,156 @@ def teardown_test_environment():
+
+def teardown_databases(old_config, verbosity, parallel=0, keepdb=False):
+    """
+    Destroys all the non-mirror databases.
```
Destroy

timgraham commented on docs/topics/testing/advanced.txt:
```diff
@@ -615,6 +609,24 @@ utility methods in the ``django.test.utils`` module.
+    function at the conclusion of testing.
+
+.. function:: teardown_databases(old_config, parallel=0, keepdb=False)
+
```
.. versionadded:: 1.11

timgraham commented:
merged in e76981b43325da60b8a7475661df6cbfa7fda37e, thanks!

pelme commented:
@timgraham Thanks for your patience and fixing up the last bits. 😃  👍
